### PR TITLE
Unquote csv strings

### PR DESCRIPTION
**Pull request by jbruechert**

### CODE REVIEW

**Comment by jbruechert** [comment]
meh, the one case that wasn't easy to solve is the relevant one for MOTIS :(

EDIT: This may also not be the right place to do this

**Comment by felixguendling** [comment]
`parse_arg` is still quite generic and also used in the HAFAS Rohdaten parser. Since the double quote escaping seems to be CSV specific, it could make sense to have it in a separate function that's only used by the CSV parser.

Maybe something overloading like this:

- `parse_csv(cstr s, std::string& t) { /* escape */ }`
- `parse_csv(cstr s, T& t) { parse_arg(cstr, t); }`

General note:
- I try to use [AAA style](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) everywhere. I didn't do it since the beginning, so you'll find plenty of places where it's not AAA style. However, for new code it makes sense to stick to it.
- I would recommend to configure a "format on save" to have code formatted in the project specific way by default (no action required). Otherwise a commit hook or manual check before committing helps.
- In general the parser code should avoid temporary and unnecessary allocations at all costs to stay fast. Your code looks fine for now though. :+1:

**Comment by felixguendling** [comment]
> meh, the one case that wasn't easy to solve is the relevant one for MOTIS :(
> 
> EDIT: This may also not be the right place to do this

I guess the way you started it here would require to change row types in these structs to `std::string`:
https://github.com/motis-project/nigiri/blob/55dde0d22013150f7db98054f336555c5003a781/src/loader/gtfs/stop.cc#L178-L179

In this case, it makes sense to `std::move` the string into the struct `stop` struct (or in other places the `trip` struct, etc.) to prevent allocating and copying more memory than necessary.

File: include/utl/parser/arg_parser.h [hunk 1]
```diff
@@ -127,11 +127,30 @@ inline bool parse_arg(cstr& s, bool& b) {
...
 }
 
-inline void parse_arg(cstr& s, std::string& arg) { arg.assign(s.str, s.len); }
+inline void parse_arg(cstr& s, std::string& arg) {
+  arg.assign(s.str, s.len);
+
+  for (int found_at = 0;
+       found_at != -1;
+       found_at = arg.find('"', found_at)) {
+    if (arg[found_at] == '"' && found_at < int(arg.size()) - 1) {
```
Comment (felixguendling) [inline include/utl/parser/arg_parser.h]: If possible (here it is) replace C-style casts with the more specific C++ versions (here: static_cast`).

Maybe the loop would be more readable with `while ((found_at = arg.find('"')) != -1) { ... }`.

**Comment by jbruechert** [comment]
> I guess the way you started it here would require to change row types in these structs to `std::string`: https://github.com/motis-project/nigiri/blob/55dde0d22013150f7db98054f336555c5003a781/src/loader/gtfs/stop.cc#L178-L179
> 
> In this case, it makes sense to `std::move` the string into the struct `stop` struct (or in other places the `trip` struct, etc.) to prevent allocating and copying more memory than necessary.

Do you think that is a good way to go? I'm a bit unhappy about the inconsistency of the overloads between cstr and std::string.

**Comment by felixguendling** [comment]
The question is: what is the alternative? To apply the unescaping function manually where necessary? But then where is it not necessary? Even for `agency_id`, etc. it's probably required if we do it correctly.

I think it's still good to have an escape hatch to prevent unescaping if you use `cstr` instead of `std::string`. The only problem I see is that it's not really obvious that one will be escaped and the other one not.

I think as long as there are no extra allocations, I am fine. I am a bit crazy with this.. especially temporary allocations :exploding_head:

**Comment by jbruechert** [comment]
Makes sense.

Now for some reason I can't get the std::string overload to be called from stop.cc, I just hope the cause are not implicit conversions to cstr, since code depends on them all over the place.

EDIT: Okay no, I just got the wrong stop.cc and didn't notice.

**Comment by jbruechert** [comment]
works now :+1:

File: include/utl/parser/csv.h [hunk 2]
```diff
@@ -71,8 +81,13 @@ inline void parse_value(cstr& s, bool& arg) {
   s = s.skip_whitespace_front();
   parse_arg(s, arg);
 }
-inline void parse_value(cstr& s, std::string& arg) { parse_arg(s, arg); }
-inline void parse_value(cstr& s, cstr& arg) { parse_arg(s, arg); }
+inline void parse_value(cstr& s, std::string& arg) {
+  parse_arg(s, arg);
+  unescape_quoted_string(arg);
+}
```
Comment (felixguendling) [inline include/utl/parser/csv.h:90]:
```suggestion
}

```

File: include/utl/parser/csv.h [hunk 3]
```diff
@@ -53,6 +53,16 @@ inline void parse_column(cstr& s, T& arg) {
                                             adjust_for_quote + adjust_for_cr));
 }
 
+inline void unescape_quoted_string(std::string &arg) {
```
Comment (felixguendling) [inline include/utl/parser/csv.h]:
```suggestion
inline void unescape_quoted_string(std::string& arg) {
```

File: include/utl/parser/csv.h [hunk 4]
```diff
@@ -53,6 +53,16 @@ inline void parse_column(cstr& s, T& arg) {
...
+  auto found_at = 0;
+  while ((found_at = arg.find('"', found_at)) != -1) {
+    if (arg[found_at] == '"' && found_at < static_cast<int>(arg.size()) - 1) {
+      arg.erase(found_at, 1);
+      found_at++; // Skip following character
```
Comment (felixguendling) [inline include/utl/parser/csv.h]:
```suggestion
      ++found_at; // Skip following character
```

Comment (felixguendling) [reply include/utl/parser/csv.h]: If both options do the same, `++i` is always preferable as it doesn't create a temporary. In `Release` mode it probably doesn't matter.

File: include/utl/parser/csv.h [hunk 5]
```diff
@@ -71,8 +81,13 @@ inline void parse_value(cstr& s, bool& arg) {
   s = s.skip_whitespace_front();
   parse_arg(s, arg);
 }
-inline void parse_value(cstr& s, std::string& arg) { parse_arg(s, arg); }
-inline void parse_value(cstr& s, cstr& arg) { parse_arg(s, arg); }
+inline void parse_value(cstr& s, std::string& arg) {
```
Comment (felixguendling) [inline include/utl/parser/csv.h:87]:
```suggestion

inline void parse_value(cstr& s, std::string& arg) {
```

**Review by felixguendling — CHANGES_REQUESTED** [review]
please run the formatter after these changes - then I think we're good to go :+1:

File: include/utl/parser/csv.h [hunk 6]
```diff
@@ -53,6 +53,16 @@ inline void parse_column(cstr& s, T& arg) {
                                             adjust_for_quote + adjust_for_cr));
 }
 
+inline void unescape_quoted_string(std::string &arg) {
+  auto found_at = 0;
+  while ((found_at = arg.find('"', found_at)) != -1) {
+    if (arg[found_at] == '"' && found_at < static_cast<int>(arg.size()) - 1) {
```
Comment (felixguendling) [inline include/utl/parser/csv.h]: `arg[found_at] == '"'` isn't this always true as the loop condition calls `found_at = arg.find('"')`?
The code replaces double quotes with single quotes, correct?
If that's true: why not use [`std::replace`](https://en.cppreference.com/w/cpp/algorithm/replace)?

Comment (jbruechert) [reply include/utl/parser/csv.h]: I don't think std::replace handle multi-element replacements.

the loop condition is missing a + 1, I'll fix it.

File: include/utl/parser/csv.h (hunk 5 above)
Comment (jbruechert) [reply include/utl/parser/csv.h:87]: This isn't new code, and the missing newlines are the same on the whole block of overloads.

File: include/utl/parser/csv.h (hunk 3 above)
Comment (jbruechert) [reply include/utl/parser/csv.h]: I've now set up clang-format in QtCreator :)

File: include/utl/parser/csv.h (hunk 6 above)
Comment (felixguendling) [reply include/utl/parser/csv.h]: Ah, right. Then maybe it's a `boost::replace_all`? Or does this solution have better performance?

Comment (jbruechert) [reply include/utl/parser/csv.h]: I haven't really benchmarked it.
Coming from the Qt world I don't know too well what's available in boost, will try replace_all.

Comment (jbruechert) [reply include/utl/parser/csv.h]: boost::replace_all seems to be based around a dequeue and probably allocates memory. Maybe my function could actually be faster.

Comment (felixguendling) [reply include/utl/parser/csv.h]: Ok, then let's stick to your version.

File: include/utl/parser/csv.h [hunk 7]
```diff
@@ -53,6 +53,17 @@ inline void parse_column(cstr& s, T& arg) {
                                             adjust_for_quote + adjust_for_cr));
 }
 
+inline void unescape_quoted_string(std::string& arg) {
+  auto found_at = 0;
+  while ((found_at = arg.find('"', found_at)) != -1) {
```
Comment (felixguendling) [inline include/utl/parser/csv.h]:
```suggestion
  while ((found_at = arg.find('"', found_at)) != std::string::npos) {
```

Not sure if it's defined to be `-1`.

File: include/utl/parser/csv.h [hunk 8]
```diff
@@ -53,6 +53,17 @@ inline void parse_column(cstr& s, T& arg) {
...
+    if (found_at < static_cast<int>(arg.size()) - 1 &&
+        arg[found_at + 1] == '"') {
+      arg.erase(found_at, 1);
+      ++found_at;  // Skip following character
```
Comment (felixguendling) [inline include/utl/parser/csv.h]: What if `found_at == arg.size() - 1`? Will it end in an infinite loop because you don't increment?

**Review by felixguendling — CHANGES_REQUESTED** [review]

**Comment by felixguendling** [comment]
Can you please add a test-case for the `found_at == s.size() - 1` case?

Somehow I don't get an email notification for force-push. So if you want me to have a look at it, it would help to request a new review or comment here. Or is there a way to subscribe to force-push events too?

**Review by felixguendling — APPROVED** [review]

**Comment by felixguendling** [comment]
Thank you very much!

**Pull request merged**